File: packages/ocp-shuffle/src/eqdlog.ts
import type { GroupElement, Scalar } from "@onchainpoker/ocp-crypto";
import {
  Transcript,
  groupElementFromBytes,
  groupElementToBytes,
  mulPoint,
  pointAdd,
  pointEq,
  pointSub,
  scalarAdd,
  scalarMul,
  scalarSub,
  scalarToBytes,
  scalarFromBytes,
} from "@onchainpoker/ocp-crypto";
import { concatBytes } from "@onchainpoker/ocp-crypto";
import { encodePoint, decodePoint, encodeScalar, decodeScalar, Reader } from "./encoding.js";
import type { ScalarRng } from "./rng.js";

export type EqDlogProof = {
  t1: GroupElement;
  t2: GroupElement;
  z: Scalar;
};

// Prove knowledge of x such that:
//   X = x*A and Y = x*B
export function proveEqDlog(params: {
  domain: string;
  A: GroupElement;
  B: GroupElement;
  X: GroupElement;
  Y: GroupElement;
  x: Scalar;
  rng: ScalarRng;
}): EqDlogProof {
  const { domain, A, B, X, Y, x, rng } = params;

  const w = rng.nextScalar();
  const t1 = mulPoint(A, w);
  const t2 = mulPoint(B, w);

  const tr = new Transcript(domain);
  tr.appendMessage("A", groupElementToBytes(A));
  tr.appendMessage("B", groupElementToBytes(B));
  tr.appendMessage("X", groupElementToBytes(X));
  tr.appendMessage("Y", groupElementToBytes(Y));
  tr.appendMessage("t1", groupElementToBytes(t1));
  tr.appendMessage("t2", groupElementToBytes(t2));
  const e = tr.challengeScalar("e");

  const z = scalarAdd(w, scalarMul(e, x));
  return { t1, t2, z };
}

export function verifyEqDlog(params: {
  domain: string;
  A: GroupElement;
  B: GroupElement;
  X: GroupElement;
  Y: GroupElement;
  proof: EqDlogProof;
}): boolean {
  const { domain, A, B, X, Y, proof } = params;
  const { t1, t2, z } = proof;

  const tr = new Transcript(domain);
  tr.appendMessage("A", groupElementToBytes(A));
  tr.appendMessage("B", groupElementToBytes(B));
  tr.appendMessage("X", groupElementToBytes(X));
  tr.appendMessage("Y", groupElementToBytes(Y));
  tr.appendMessage("t1", groupElementToBytes(t1));
  tr.appendMessage("t2", groupElementToBytes(t2));
  const e = tr.challengeScalar("e");

  // Check: z*A == t1 + e*X
  const lhs1 = mulPoint(A, z);
  const rhs1 = pointAdd(t1, mulPoint(X, e));
  if (!pointEq(lhs1, rhs1)) return false;

  // Check: z*B == t2 + e*Y
  const lhs2 = mulPoint(B, z);
  const rhs2 = pointAdd(t2, mulPoint(Y, e));
  if (!pointEq(lhs2, rhs2)) return false;

  return true;
}

// Simulation helper: given chosen (e,z), compute commitments that satisfy verification equations.
export function simulateEqDlogCommitments(params: {
  A: GroupElement;
  B: GroupElement;
  X: GroupElement;
  Y: GroupElement;
  e: Scalar;
  z: Scalar;
}): { t1: GroupElement; t2: GroupElement } {
  const { A, B, X, Y, e, z } = params;
  const t1 = pointSub(mulPoint(A, z), mulPoint(X, e));
  const t2 = pointSub(mulPoint(B, z), mulPoint(Y, e));
  return { t1, t2 };
}

// Encoding: t1(32) || t2(32) || z(32)
export function encodeEqDlogProof(p: EqDlogProof): Uint8Array {
  return concatBytes(groupElementToBytes(p.t1), groupElementToBytes(p.t2), scalarToBytes(p.z));
}

export function decodeEqDlogProof(bytes: Uint8Array): EqDlogProof {
  if (!(bytes instanceof Uint8Array) || bytes.length !== 96) throw new Error("decodeEqDlogProof: expected 96 bytes");
  const t1 = groupElementFromBytes(bytes.subarray(0, 32));
  const t2 = groupElementFromBytes(bytes.subarray(32, 64));
  const z = scalarFromBytes(bytes.subarray(64, 96));
  return { t1, t2, z };
}

export function decodeEqDlogProofFromReader(reader: Reader): EqDlogProof {
  return {
    t1: decodePoint(reader.take(32)),
    t2: decodePoint(reader.take(32)),
    z: decodeScalar(reader.take(32)),
  };
}

export function encodeEqDlogProofV1(p: EqDlogProof): Uint8Array {
  // Uses encoding.ts wrappers; identical to encodeEqDlogProof but avoids deep crypto imports if needed.
  return concatBytes(encodePoint(p.t1), encodePoint(p.t2), encodeScalar(p.z));
}

File: apps/cosmos/internal/ocpshuffle/eqdlog.go
package ocpshuffle

import (
	"fmt"

	"onchainpoker/apps/cosmos/internal/ocpcrypto"
)

type EqDlogProof struct {
	T1 ocpcrypto.Point
	T2 ocpcrypto.Point
	Z  ocpcrypto.Scalar
}

// Prove knowledge of x such that:
//   X = x*A and Y = x*B
func proveEqDlog(domain string, A ocpcrypto.Point, B ocpcrypto.Point, X ocpcrypto.Point, Y ocpcrypto.Point, x ocpcrypto.Scalar, rng scalarRng) (EqDlogProof, error) {
	w, err := rng.NextScalar()
	if err != nil {
		return EqDlogProof{}, err
	}
	t1 := ocpcrypto.MulPoint(A, w)
	t2 := ocpcrypto.MulPoint(B, w)

	tr := ocpcrypto.NewTranscript(domain)
	_ = tr.AppendMessage("A", A.Bytes())
	_ = tr.AppendMessage("B", B.Bytes())
	_ = tr.AppendMessage("X", X.Bytes())
	_ = tr.AppendMessage("Y", Y.Bytes())
	_ = tr.AppendMessage("t1", t1.Bytes())
	_ = tr.AppendMessage("t2", t2.Bytes())
	e, err := tr.ChallengeScalar("e")
	if err != nil {
		return EqDlogProof{}, err
	}

	z := ocpcrypto.ScalarAdd(w, ocpcrypto.ScalarMul(e, x))
	return EqDlogProof{T1: t1, T2: t2, Z: z}, nil
}

func verifyEqDlog(domain string, A ocpcrypto.Point, B ocpcrypto.Point, X ocpcrypto.Point, Y ocpcrypto.Point, proof EqDlogProof) (bool, error) {
	tr := ocpcrypto.NewTranscript(domain)
	_ = tr.AppendMessage("A", A.Bytes())
	_ = tr.AppendMessage("B", B.Bytes())
	_ = tr.AppendMessage("X", X.Bytes())
	_ = tr.AppendMessage("Y", Y.Bytes())
	_ = tr.AppendMessage("t1", proof.T1.Bytes())
	_ = tr.AppendMessage("t2", proof.T2.Bytes())
	e, err := tr.ChallengeScalar("e")
	if err != nil {
		return false, err
	}

	// Check: z*A == t1 + e*X
	lhs1 := ocpcrypto.MulPoint(A, proof.Z)
	rhs1 := ocpcrypto.PointAdd(proof.T1, ocpcrypto.MulPoint(X, e))
	if !ocpcrypto.PointEq(lhs1, rhs1) {
		return false, nil
	}

	// Check: z*B == t2 + e*Y
	lhs2 := ocpcrypto.MulPoint(B, proof.Z)
	rhs2 := ocpcrypto.PointAdd(proof.T2, ocpcrypto.MulPoint(Y, e))
	if !ocpcrypto.PointEq(lhs2, rhs2) {
		return false, nil
	}
	return true, nil
}

// Simulation helper: given chosen (e,z), compute commitments that satisfy verification equations.
func simulateEqDlogCommitments(A ocpcrypto.Point, B ocpcrypto.Point, X ocpcrypto.Point, Y ocpcrypto.Point, e ocpcrypto.Scalar, z ocpcrypto.Scalar) (ocpcrypto.Point, ocpcrypto.Point) {
	t1 := ocpcrypto.PointSub(ocpcrypto.MulPoint(A, z), ocpcrypto.MulPoint(X, e))
	t2 := ocpcrypto.PointSub(ocpcrypto.MulPoint(B, z), ocpcrypto.MulPoint(Y, e))
	return t1, t2
}

func encodeEqDlogProof(p EqDlogProof) []byte {
	return append(append(p.T1.Bytes(), p.T2.Bytes()...), p.Z.Bytes()...)
}

func decodeEqDlogProofFromReader(r *reader) (EqDlogProof, error) {
	t1b, err := r.take(32)
	if err != nil {
		return EqDlogProof{}, err
	}
	t2b, err := r.take(32)
	if err != nil {
		return EqDlogProof{}, err
	}
	zb, err := r.take(32)
	if err != nil {
		return EqDlogProof{}, err
	}
	t1, err := decodePoint(t1b)
	if err != nil {
		return EqDlogProof{}, err
	}
	t2, err := decodePoint(t2b)
	if err != nil {
		return EqDlogProof{}, err
	}
	z, err := decodeScalar(zb)
	if err != nil {
		return EqDlogProof{}, err
	}
	return EqDlogProof{T1: t1, T2: t2, Z: z}, nil
}

func decodeEqDlogProof(bytes []byte) (EqDlogProof, error) {
	if len(bytes) != 96 {
		return EqDlogProof{}, fmt.Errorf("decodeEqDlogProof: expected 96 bytes")
	}
	r := newReader(bytes)
	p, err := decodeEqDlogProofFromReader(r)
	if err != nil {
		return EqDlogProof{}, err
	}
	if !r.done() {
		return EqDlogProof{}, fmt.Errorf("decodeEqDlogProof: trailing bytes")
	}
	return p, nil
}

